fe/java-udf/src/main/java/org/apache/doris/jni/vec/VectorColumn.java

```diff
@@ -0,0 +1,573 @@
+
+    public VectorColumn(ColumnType columnType, int capacity) {
+        this.columnType = columnType;
+        this.capacity = 0;
+        this.nullMap = 0;
+        this.data = 0;
+        this.offsets = 0;
+        this.numNulls = 0;
+        this.appendIndex = 0;
+        if (columnType.isComplexType()) {
+            List<ColumnType> children = columnType.getChildTypes();
+            childColumns = new VectorColumn[children.size()];
+            for (int i = 0; i < children.size(); ++i) {
+                childColumns[i] = new VectorColumn(children.get(i), capacity);
+            }
+        } else if (columnType.isStringType()) {
+            childColumns = new VectorColumn[1];
+            childColumns[0] = new VectorColumn(new ColumnType("#data", Type.BYTE), capacity * DEFAULT_STRING_LENGTH);
+        }
+
+        reserveCapacity(capacity);
+    }
+
+    public long nullMapAddress() {
+        return nullMap;
+    }
+
+    public long dataAddress() {
+        return data;
+    }
+
+    public long offsetAddress() {
+        return offsets;
+    }
+
+    /**
+     * Release columns and meta information
+     */
+    public void close() {
+        if (childColumns != null) {
+            for (int i = 0; i < childColumns.length; i++) {
+                childColumns[i].close();
+                childColumns[i] = null;
+            }
+            childColumns = null;
+        }
+
+        if (nullMap != 0) {
+            OffHeap.freeMemory(nullMap);
+        }
+        if (data != 0) {
+            OffHeap.freeMemory(data);
+        }
+        if (offsets != 0) {
+            OffHeap.freeMemory(offsets);
+        }
+        nullMap = 0;
+        data = 0;
+        offsets = 0;
+        capacity = 0;
+        numNulls = 0;
+        appendIndex = 0;
+    }
+
+    private void throwReserveException(int requiredCapacity, Throwable cause) {
+        String message = "Cannot reserve enough bytes in off heap memory ("
+                + (requiredCapacity >= 0 ? "requested " + requiredCapacity + " bytes" : "integer overflow).");
+        throw new RuntimeException(message, cause);
+    }
+
+    private void reserve(int requiredCapacity) {
+        if (requiredCapacity < 0) {
+            throwReserveException(requiredCapacity, null);
+        } else if (requiredCapacity > capacity) {
+            int newCapacity = (int) Math.min(MAX_CAPACITY, requiredCapacity * 2L);
+            if (requiredCapacity <= newCapacity) {
+                try {
+                    reserveCapacity(newCapacity);
+                } catch (OutOfMemoryError outOfMemoryError) {
+                    throwReserveException(requiredCapacity, outOfMemoryError);
+                }
+            } else {
+                // overflow
+                throwReserveException(requiredCapacity, null);
+            }
+        }
+    }
+
+    private void reserveCapacity(int newCapacity) {
+        long oldCapacity = capacity;
+        long oldOffsetSize = capacity * 4L;
```

Comment:
Why multiply 4?

Reply:
`capacity` is the number of elements in vector column, and `size` is the memory usage in vector column. We use `int` to store offsets, so the size of offsets should multiply 4.